src/encoder.rs

```diff
@@ -104,6 +109,17 @@ pub struct Writer<W: Write> {
     info: Info,
 }
 
+fn write_chunk<W: Write>(w: &mut W, name: [u8; 4], data: &[u8]) -> Result<()> {
```

Comment:
Is there a reason for using this function instead of `Writer::write_chunk`?

Reply:
It seems already used in `Writer::new` but the method requires `&self`. The freestanding function however does not.

Reply:
Yeah, but this function was used in `init` which takes `&self`.

Reply:
Oh stupid me. Github's doesn't actual give the right context. I thought line 135 was still within `new`.

Reply:
No, sorry that was wrong, `init` takes `mut self`, but the function should still be usable.

Reply:
I struggled to both borrow mutably self and immutably its palette. That's why I split it up so that I only had to borrow mutably the writer.

Reply:
Ah I see. Well, one standard trick you can employ is to `take` from an `Option` in that case:
```
// Temporarily take remove it from `self`.
if let Some(palette) = self.info.palette.take() {
    self.write_chunk(chunk::PLTE, &palette)?;
    // Put it back in, we check that later.
    self.info.palette = Some(palette)
}
```

Reply:
But there's nothing wrong with having it as a separate method, actually. I'd even go so far and say it is nicer? It was only a bit confusing to have it included in this feature addition and not as part of a code polish.

Reply:
The only thing I'd change is the exact function definition:
```
// `w: W` instead of `w: &mut W`
fn write_chunk<W: Write>(w: W, name: [u8; 4], data: &[u8]) -> Result<()> {
```
That is because `Write` is automatically implemented for `&'_ mut impl Write` and so this is slightly more generic.

Reply:
Yes, I have nothing against it. I only wondered why it ended up like that.

Reply:
It won't accept that:
```
error[E0596]: cannot borrow `w` as mutable, as it is not declared as mutable
   --> src/encoder.rs:113:5
    |
112 | fn write_chunk<W: Write>(w: W, name: [u8; 4], data: &[u8]) -> Result<()> {
    |                          - help: consider changing this to be mutable: `mut w`
113 |     w.write_be(data.len() as u32)?;
    |     ^ cannot borrow as mutable
```

Reply:
`fn ...(mut w: W, ` since you use it mutably within the function, yeah

Reply:
Oh cool – learning something here :-)